mono/metadata/threads.c

```diff
@@ -1637,57 +1637,91 @@ ves_icall_System_Threading_Thread_GetName_internal (MonoInternalThread *this_obj
+	const char* utf8, const gunichar2 *utf16, gboolean permanent,
+	gboolean reset, MonoError *error)
 {
+	error_init (error);
+
+	// If length is set, then utf16 is required.
+	g_assert (!length || utf16);
+
+	gunichar2* dup16 = NULL;
+	gunichar2* old_name = NULL;
+	char* dup8 = NULL;
 	MonoNativeThreadId tid = 0;
+	gboolean unlock = FALSE;
 
-	LOCK_THREAD (this_obj);
+	// Duplicate utf16 outside lock.
+	if (length > 0)
+		dup16 = (gunichar2*)g_memdup (utf16, length * sizeof (gunichar2));
 
-	error_init (error);
+	LOCK_THREAD (this_obj);
+	unlock = TRUE;
 
 	if (reset) {
 		this_obj->flags &= ~MONO_THREAD_FLAG_NAME_SET;
 	} else if (this_obj->flags & MONO_THREAD_FLAG_NAME_SET) {
 		UNLOCK_THREAD (this_obj);
-		
+		unlock = FALSE;
 		mono_error_set_invalid_operation (error, "Thread.Name can only be set once.");
-		return;
-	}
-	if (this_obj->name) {
-		g_free (this_obj->name);
-		this_obj->name_len = 0;
+		goto Exit;
 	}
-	if (name) {
-		this_obj->name = g_memdup (mono_string_chars (name), mono_string_length (name) * sizeof (gunichar2));
-		this_obj->name_len = mono_string_length (name);
 
+	old_name = this_obj->name;
+	this_obj->name = NULL;
+	this_obj->name_len = 0;
+
+	if (length > 0) {
+		this_obj->name = dup16;
+		dup16 = NULL;
+		this_obj->name_len = length;
 		if (permanent)
 			this_obj->flags |= MONO_THREAD_FLAG_NAME_SET;
 	}
-	else
-		this_obj->name = NULL;
 
 	if (!(this_obj->state & ThreadState_Stopped))
 		tid = thread_get_tid (this_obj);
 
 	UNLOCK_THREAD (this_obj);
+	unlock = FALSE;
 
 	if (this_obj->name && tid) {
-		char *tname = mono_string_to_utf8_checked (name, error);
-		return_if_nok (error);
-		MONO_PROFILER_RAISE (thread_name, ((uintptr_t)tid, tname));
-		mono_native_thread_set_name (tid, tname);
-		mono_free (tname);
+		const char *profiler_name = NULL;
+		if (utf8)
+			profiler_name = utf8;
+		else {
+			dup8 = mono_utf16_to_utf8 (utf16, length, error);
+			goto_if_nok (error, Exit);
+			profiler_name = dup8;
+		}
+		MONO_PROFILER_RAISE (thread_name, ((uintptr_t)tid, profiler_name));
+		mono_native_thread_set_name (tid, profiler_name);
 	}
+Exit:
+	if (unlock)
+		UNLOCK_THREAD (this_obj);
+	g_free (old_name);
+	g_free (dup8);
+	g_free (dup16);
 }
 
 void 
-ves_icall_System_Threading_Thread_SetName_internal (MonoInternalThread *this_obj, MonoString *name)
+ves_icall_System_Threading_Thread_SetName_internal (
+	MonoInternalThreadHandle thread, MonoStringHandle name_handle, MonoError *error)
 {
-	MonoError error;
-	mono_thread_set_name_internal (this_obj, name, TRUE, FALSE, &error);
-	mono_error_set_pending_exception (&error);
+	error_init (error);
+
+	MONO_CHECK_ARG_NULL (thread,);
+	guint const thread_gchandle = mono_gchandle_from_handle (MONO_HANDLE_CAST (MonoObject, thread), TRUE);
+
+	mono_unwrapped_string_t name = mono_unwrap_string_handle (name_handle);
+
+	mono_thread_set_name_internal (MONO_HANDLE_RAW (thread), name.length, NULL, name.chars, TRUE, FALSE, error);
```

Comment:
Is this line actually sufficient? Or need to pass handle mono_thread_set_name_internal?

Reply:
I think it is sufficient, since since internal threads are pinned. I don't know if we want to have that/depend on that long term.